src/ObjCRuntime/Runtime.CoreCLR.cs

```diff
@@ -182,6 +189,183 @@ static bool IsInstance (MonoObjectPtr mobj, MonoObjectPtr mtype)
+
+			var rv = method.Invoke (instance, BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Static | BindingFlags.Instance, null, parameters, null);
+
+			// Copy any byref parameters back out again
+			var byrefParameterCount = 0;
+			for (var i = 0; i < methodParameters.Length; i++) {
+				var p = methodParameters [i];
+				if (!p.IsOut && !p.ParameterType.IsByRef)
+					continue;
+
+				byrefParameterCount++;
+
+				log_coreclr ($"    Marshalling #{i + 1} back (Type: {p.ParameterType.FullName}) value: {(parameters [i] == null ? "<null>" : parameters [i].GetType ().FullName)}");
+
+				var parameterType = p.ParameterType.GetElementType ();
+				var isMonoObject = parameterType.IsClass || parameterType.IsInterface || (parameterType.IsValueType && IsNullable (parameterType));
+
+				var nativeParam = nativeParameters [i];
+
+				if (nativeParam == IntPtr.Zero) {
+					log_coreclr ($"    No output pointer was provided.");
+					continue;
+				}
+
+				if (parameters [i] == inputParameters [i]) {
+					log_coreclr ($"        The argument didn't change, no marshalling required");
+					if (parameters [i] != null && parameterType != typeof (IntPtr) && isMonoObject) {
+						// byref parameters must be retained
+						xamarin_mono_object_retain (ref nativeParam);
+					}
+					continue;
+				}
+
+				if (parameterType == typeof (IntPtr)) {
+					Marshal.WriteIntPtr (nativeParam, (IntPtr) parameters [i]);
+					log_coreclr ($"        IntPtr: 0x{((IntPtr) parameters [i]).ToString ("x")} => Type: {parameters [i]?.GetType ()} nativeParam: 0x{nativeParam.ToString ("x")}");
+				} else if (isMonoObject) {
+					var ptr = GetMonoObject (parameters [i]);
+					Marshal.WriteIntPtr (nativeParam, ptr);
+					log_coreclr ($"        IsClass/IsInterface/IsNullable: {(parameters [i] == null ? "<null>" : parameters [i].GetType ().FullName)}  nativeParam: 0x{nativeParam.ToString ("x")} -> MonoObject: 0x{ptr.ToString ("x")}");
+				} else if (parameterType.IsValueType) {
+					StructureToPtr (parameters [i], nativeParam);
+					log_coreclr ($"        IsValueType: {(parameters [i] == null ? "<null>" : parameters [i].ToString ())} nativeParam: 0x{nativeParam.ToString ("x")}");
+				} else {
+					throw ErrorHelper.CreateError (8038, Errors.MX8038 /* Don't know how to marshal back the parameter of type {p.ParameterType.FullName} for parameter {p.Name} in call to {method} */, p.ParameterType.FullName, p.Name, method);
+				}
+			}
+
+			// we're done!
+			log_coreclr ($"    Invoke complete with {byrefParameterCount} ref parameters and return value of type {rv?.GetType ()}");
+
+			return rv;
+		}
+
+		static bool IsNullable (Type type)
+		{
+			if (Nullable.GetUnderlyingType (type) != null)
+				return true;
+
+			if (type.IsGenericType && type.GetGenericTypeDefinition () == typeof (Nullable<>))
+				return true;
+
+			return false;
+		}
+
+		static object PtrToStructure (IntPtr ptr, Type type)
+		{
+			if (ptr == IntPtr.Zero)
+				return null;
```

Comment:
required ? the API returns `object?` and the exception is not documented

Reply:
```
csharp> Marshal.PtrToStructure (IntPtr.Zero, 1);
System.ArgumentNullException: Value cannot be null.  
Parameter name: src                                  
  at (wrapper managed-to-native) System.Runtime.InteropServices.Marshal.PtrToStructure(intptr,object)
  at System.Runtime.InteropServices.Marshal.PtrToStructure[T] (System.IntPtr ptr, T structure) [0x00000] in <996c1f36e5dc440297a28fe1d8dc8fe4>:0 
```

The API returns null if given a null type:

```
csharp> Marshal.PtrToStructure (IntPtr.Zero, null);  
null   
```